// redux-saga-requests/lib/constants.js
"use strict";

exports.__esModule = true;
exports.INCORRECT_PAYLOAD_ERROR = exports.CLEAR_REQUESTS_CACHE = exports.GET_REQUEST_CACHE = exports.INTERCEPTORS = exports.RUN_BY_INTERCEPTOR = exports.REQUESTS_CONFIG = exports.ABORT_SUFFIX = exports.ERROR_SUFFIX = exports.SUCCESS_SUFFIX = void 0;
var SUCCESS_SUFFIX = '_SUCCESS';
exports.SUCCESS_SUFFIX = SUCCESS_SUFFIX;
var ERROR_SUFFIX = '_ERROR';
exports.ERROR_SUFFIX = ERROR_SUFFIX;
var ABORT_SUFFIX = '_ABORT';
exports.ABORT_SUFFIX = ABORT_SUFFIX;
var REQUESTS_CONFIG = 'REDUX_SAGA_REQUESTS_CONFIG';
exports.REQUESTS_CONFIG = REQUESTS_CONFIG;
var RUN_BY_INTERCEPTOR = 'RUN_BY_INTERCEPTOR';
exports.RUN_BY_INTERCEPTOR = RUN_BY_INTERCEPTOR;
var INTERCEPTORS = {
  ON_REQUEST: 'onRequest',
  ON_ERROR: 'onError',
  ON_SUCCESS: 'onSuccess',
  ON_ABORT: 'onAbort'
};
exports.INTERCEPTORS = INTERCEPTORS;
var GET_REQUEST_CACHE = 'GET_REQUEST_CACHE';
exports.GET_REQUEST_CACHE = GET_REQUEST_CACHE;
var CLEAR_REQUESTS_CACHE = 'CLEAR_REQUESTS_CACHE';
exports.CLEAR_REQUESTS_CACHE = CLEAR_REQUESTS_CACHE;
var INCORRECT_PAYLOAD_ERROR = "Incorrect payload for request action. Action must have form of { type: 'TYPE', request: {} }, { type: 'TYPE', request: [{}, {}] }, { type: 'TYPE', payload: { request: {} } } or { type: 'TYPE', payload: { request: [{}, {}] } }";
exports.INCORRECT_PAYLOAD_ERROR = INCORRECT_PAYLOAD_ERROR;
