// redux-saga-requests/lib/sagas.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

exports.__esModule = true;
exports.createRequestInstance = createRequestInstance;
exports.getRequestsConfig = getRequestsConfig;
exports.getRequestInstance = getRequestInstance;
exports.sendRequest = sendRequest;
exports.cancelSendRequestOnAction = cancelSendRequestOnAction;
exports.watchRequests = watchRequests;
exports.countServerRequests = countServerRequests;
exports.defaultConfig = exports.voidCallback = void 0;

var _regenerator = _interopRequireDefault(require("@babel/runtime/regenerator"));

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _effects = require("redux-saga/effects");

var _reduxSaga = require("redux-saga");

var _actions = require("./actions");

var _constants = require("./constants");

var _marked =
/*#__PURE__*/
_regenerator.default.mark(getRequestInstance),
    _marked2 =
/*#__PURE__*/
_regenerator.default.mark(sendRequest),
    _marked3 =
/*#__PURE__*/
_regenerator.default.mark(cancelSendRequestOnAction),
    _marked4 =
/*#__PURE__*/
_regenerator.default.mark(watchRequests),
    _marked5 =
/*#__PURE__*/
_regenerator.default.mark(countServerRequests);

/* eslint-disable */
var delay = require('redux-saga').delay || require('@redux-saga/delay-p').default;
/* eslint-enable */


var voidCallback = function voidCallback() {};

exports.voidCallback = voidCallback;
var defaultConfig = {
  driver: null,
  onRequest: null,
  onSuccess: null,
  onError: null,
  onAbort: null
};
exports.defaultConfig = defaultConfig;

function createRequestInstance(config) {
  var _setContext;

  return (0, _effects.setContext)((_setContext = {}, _setContext[_constants.REQUESTS_CONFIG] = (0, _extends2.default)({}, defaultConfig, config), _setContext));
}

function getRequestsConfig() {
  return (0, _effects.getContext)(_constants.REQUESTS_CONFIG);
}

function getRequestInstance(driverType) {
  var config, driver;
  return _regenerator.default.wrap(function getRequestInstance$(_context) {
    while (1) {
      switch (_context.prev = _context.next) {
        case 0:
          if (driverType === void 0) {
            driverType = null;
          }

          _context.next = 3;
          return getRequestsConfig();

        case 3:
          config = _context.sent;
          driver = driverType ? config.driver[driverType] : config.driver.default || config.driver;
          return _context.abrupt("return", driver.requestInstance);

        case 6:
        case "end":
          return _context.stop();
      }
    }
  }, _marked, this);
}

var getDriver = function getDriver(requestsConfig, action) {
  return action.meta && action.meta.driver ? requestsConfig.driver[action.meta.driver] : requestsConfig.driver.default || requestsConfig.driver;
};

function sendRequest(action, _temp) {
  var _ref, _ref$dispatchRequestA, dispatchRequestAction, _ref$silent, silent, _ref$runOnRequest, runOnRequest, _ref$runOnSuccess, runOnSuccess, _ref$runOnError, runOnError, _ref$runOnAbort, runOnAbort, runByInterceptor, requestsConfig, driver, actionPayload, _setContext2, abortSource, response, responseError, _setContext3, _ref2, onErrorResponse, onErrorError, errorPayload, _setContext4, successPayload, _setContext5;

  return _regenerator.default.wrap(function sendRequest$(_context2) {
    while (1) {
      switch (_context2.prev = _context2.next) {
        case 0:
          _ref = _temp === void 0 ? {} : _temp, _ref$dispatchRequestA = _ref.dispatchRequestAction, dispatchRequestAction = _ref$dispatchRequestA === void 0 ? false : _ref$dispatchRequestA, _ref$silent = _ref.silent, silent = _ref$silent === void 0 ? false : _ref$silent, _ref$runOnRequest = _ref.runOnRequest, runOnRequest = _ref$runOnRequest === void 0 ? null : _ref$runOnRequest, _ref$runOnSuccess = _ref.runOnSuccess, runOnSuccess = _ref$runOnSuccess === void 0 ? null : _ref$runOnSuccess, _ref$runOnError = _ref.runOnError, runOnError = _ref$runOnError === void 0 ? null : _ref$runOnError, _ref$runOnAbort = _ref.runOnAbort, runOnAbort = _ref$runOnAbort === void 0 ? null : _ref$runOnAbort;

          if ((0, _actions.isRequestAction)(action)) {
            _context2.next = 3;
            break;
          }

          throw new Error(_constants.INCORRECT_PAYLOAD_ERROR);

        case 3:
          _context2.next = 5;
          return (0, _effects.getContext)(_constants.RUN_BY_INTERCEPTOR);

        case 5:
          runByInterceptor = _context2.sent;
          _context2.next = 8;
          return getRequestsConfig();

        case 8:
          requestsConfig = _context2.sent;

          if (!(dispatchRequestAction && !silent)) {
            _context2.next = 16;
            break;
          }

          action = (0, _extends2.default)({}, action, {
            meta: (0, _extends2.default)({}, action.meta, {
              runByWatcher: false
            })
          });
          _context2.next = 13;
          return (0, _effects.put)(action);

        case 13:
          action = _context2.sent;

          if (!(action === null)) {
            _context2.next = 16;
            break;
          }

          return _context2.abrupt("return", {
            serverSide: true
          });

        case 16:
          _context2.next = 18;
          return (0, _effects.call)(getDriver, requestsConfig, action);

        case 18:
          driver = _context2.sent;
          actionPayload = (0, _actions.getActionPayload)(action);

          if (!(requestsConfig.onRequest && (runOnRequest !== null ? runOnRequest : runByInterceptor !== _constants.INTERCEPTORS.ON_REQUEST))) {
            _context2.next = 26;
            break;
          }

          _context2.next = 23;
          return (0, _effects.setContext)((_setContext2 = {}, _setContext2[_constants.RUN_BY_INTERCEPTOR] = _constants.INTERCEPTORS.ON_REQUEST, _setContext2));

        case 23:
          _context2.next = 25;
          return (0, _effects.call)(requestsConfig.onRequest, actionPayload.request, action);

        case 25:
          actionPayload.request = _context2.sent;

        case 26:
          abortSource = driver.getAbortSource();
          _context2.prev = 27;
          _context2.prev = 28;

          if (!(action.meta && action.meta.cacheResponse)) {
            _context2.next = 33;
            break;
          }

          response = action.meta.cacheResponse;
          _context2.next = 42;
          break;

        case 33:
          if (Array.isArray(actionPayload.request)) {
            _context2.next = 39;
            break;
          }

          _context2.next = 36;
          return (0, _effects.call)([driver, 'sendRequest'], actionPayload.request, abortSource, action);

        case 36:
          response = _context2.sent;
          _context2.next = 42;
          break;

        case 39:
          _context2.next = 41;
          return (0, _effects.all)(actionPayload.request.map(function (requestItem) {
            return (0, _effects.call)([driver, 'sendRequest'], requestItem, abortSource, action);
          }));

        case 41:
          response = _context2.sent;

        case 42:
          _context2.next = 47;
          break;

        case 44:
          _context2.prev = 44;
          _context2.t0 = _context2["catch"](28);
          responseError = _context2.t0;

        case 47:
          if (!responseError) {
            _context2.next = 65;
            break;
          }

          if (!(requestsConfig.onError && (runOnError !== null ? runOnError : runByInterceptor !== _constants.INTERCEPTORS.ON_ERROR))) {
            _context2.next = 57;
            break;
          }

          _context2.next = 51;
          return (0, _effects.setContext)((_setContext3 = {}, _setContext3[_constants.RUN_BY_INTERCEPTOR] = _constants.INTERCEPTORS.ON_ERROR, _setContext3));

        case 51:
          _context2.next = 53;
          return (0, _effects.call)(requestsConfig.onError, responseError, action);

        case 53:
          _ref2 = _context2.sent;
          onErrorResponse = _ref2.response;
          onErrorError = _ref2.error;

          if (onErrorResponse) {
            response = onErrorResponse;
          } else {
            responseError = onErrorError;
          }

        case 57:
          if (response) {
            _context2.next = 65;
            break;
          }

          _context2.next = 60;
          return (0, _effects.call)(driver.getErrorPayload, responseError);

        case 60:
          errorPayload = _context2.sent;

          if (silent) {
            _context2.next = 64;
            break;
          }

          _context2.next = 64;
          return (0, _effects.put)((0, _actions.createErrorAction)(action, errorPayload));

        case 64:
          return _context2.abrupt("return", {
            error: responseError
          });

        case 65:
          if (!(requestsConfig.onSuccess && (runOnSuccess !== null ? runOnSuccess : runByInterceptor !== _constants.INTERCEPTORS.ON_SUCCESS && runByInterceptor !== _constants.INTERCEPTORS.ON_ERROR))) {
            _context2.next = 71;
            break;
          }

          _context2.next = 68;
          return (0, _effects.setContext)((_setContext4 = {}, _setContext4[_constants.RUN_BY_INTERCEPTOR] = _constants.INTERCEPTORS.ON_SUCCESS, _setContext4));

        case 68:
          _context2.next = 70;
          return (0, _effects.call)(requestsConfig.onSuccess, response, action);

        case 70:
          response = _context2.sent;

        case 71:
          _context2.next = 73;
          return (0, _effects.call)(driver.getSuccessPayload, response, actionPayload.request);

        case 73:
          successPayload = _context2.sent;

          if (silent) {
            _context2.next = 77;
            break;
          }

          _context2.next = 77;
          return (0, _effects.put)((0, _actions.createSuccessAction)(action, successPayload, response));

        case 77:
          return _context2.abrupt("return", {
            response: response
          });

        case 78:
          _context2.prev = 78;
          _context2.next = 81;
          return (0, _effects.cancelled)();

        case 81:
          if (!_context2.sent) {
            _context2.next = 92;
            break;
          }

          _context2.next = 84;
          return (0, _effects.call)([driver, 'abortRequest'], abortSource);

        case 84:
          if (!(requestsConfig.onAbort && (runOnAbort !== null ? runOnAbort : runByInterceptor !== _constants.INTERCEPTORS.ON_ABORT))) {
            _context2.next = 89;
            break;
          }

          _context2.next = 87;
          return (0, _effects.setContext)((_setContext5 = {}, _setContext5[_constants.RUN_BY_INTERCEPTOR] = _constants.INTERCEPTORS.ON_ABORT, _setContext5));

        case 87:
          _context2.next = 89;
          return (0, _effects.call)(requestsConfig.onAbort, action);

        case 89:
          if (silent) {
            _context2.next = 92;
            break;
          }

          _context2.next = 92;
          return (0, _effects.put)((0, _actions.createAbortAction)(action));

        case 92:
          return _context2.finish(78);

        case 93:
        case "end":
          return _context2.stop();
      }
    }
  }, _marked2, this, [[27,, 78, 93], [28, 44]]);
}

var isRequestReadOnly = function isRequestReadOnly(request) {
  return !request.query && (!request.method || request.method.toLowerCase() === 'get') || request.query && !request.query.trim().startsWith('mutation');
};

var watchRequestsDefaultConfig = {
  takeLatest: function takeLatest(action) {
    var _getActionPayload = (0, _actions.getActionPayload)(action),
        request = _getActionPayload.request;

    return Array.isArray(request) ? request.every(isRequestReadOnly) : isRequestReadOnly(request);
  },
  abortOn: null,
  getLastActionKey: function getLastActionKey(action) {
    return action.type;
  }
};

function cancelSendRequestOnAction(abortOn, task) {
  var _ref3, abortingAction;

  return _regenerator.default.wrap(function cancelSendRequestOnAction$(_context3) {
    while (1) {
      switch (_context3.prev = _context3.next) {
        case 0:
          _context3.next = 2;
          return (0, _effects.race)({
            abortingAction: (0, _effects.take)(abortOn),
            taskFinished: (0, _effects.join)(task),
            timeout: (0, _effects.call)(delay, 10000) // taskFinished doesnt work for aborted tasks

          });

        case 2:
          _ref3 = _context3.sent;
          abortingAction = _ref3.abortingAction;

          if (!abortingAction) {
            _context3.next = 7;
            break;
          }

          _context3.next = 7;
          return (0, _effects.cancel)(task);

        case 7:
        case "end":
          return _context3.stop();
      }
    }
  }, _marked3, this);
}

var isWatchable = function isWatchable(a) {
  return (0, _actions.isRequestAction)(a) && (!a.meta || a.meta.runByWatcher !== false);
};

function watchRequests(commonConfig) {
  var lastTasks, config, action, lastActionKey, takeLatest, activeTask, newTask, abortOn;
  return _regenerator.default.wrap(function watchRequests$(_context4) {
    while (1) {
      switch (_context4.prev = _context4.next) {
        case 0:
          if (commonConfig === void 0) {
            commonConfig = {};
          }

          lastTasks = {};
          config = (0, _extends2.default)({}, watchRequestsDefaultConfig, commonConfig);

        case 3:
          if (!true) {
            _context4.next = 24;
            break;
          }

          _context4.next = 6;
          return (0, _effects.take)(isWatchable);

        case 6:
          action = _context4.sent;
          lastActionKey = config.getLastActionKey(action);
          takeLatest = action.meta && action.meta.takeLatest !== undefined ? action.meta.takeLatest : typeof config.takeLatest === 'function' ? config.takeLatest(action) : config.takeLatest;

          if (!takeLatest) {
            _context4.next = 14;
            break;
          }

          activeTask = lastTasks[lastActionKey];

          if (!activeTask) {
            _context4.next = 14;
            break;
          }

          _context4.next = 14;
          return (0, _effects.cancel)(activeTask);

        case 14:
          _context4.next = 16;
          return (0, _effects.fork)(sendRequest, action);

        case 16:
          newTask = _context4.sent;

          if (takeLatest) {
            lastTasks[lastActionKey] = newTask;
          }

          abortOn = action.meta && action.meta.abortOn ? action.meta.abortOn : config.abortOn;

          if (!abortOn) {
            _context4.next = 22;
            break;
          }

          _context4.next = 22;
          return (0, _effects.fork)(cancelSendRequestOnAction, abortOn, newTask);

        case 22:
          _context4.next = 3;
          break;

        case 24:
        case "end":
          return _context4.stop();
      }
    }
  }, _marked4, this);
}

function countServerRequests(_ref4) {
  var serverRequestActions, _ref4$finishOnFirstEr, finishOnFirstError, index, action;

  return _regenerator.default.wrap(function countServerRequests$(_context5) {
    while (1) {
      switch (_context5.prev = _context5.next) {
        case 0:
          serverRequestActions = _ref4.serverRequestActions, _ref4$finishOnFirstEr = _ref4.finishOnFirstError, finishOnFirstError = _ref4$finishOnFirstEr === void 0 ? true : _ref4$finishOnFirstEr;
          index = 0;
          serverRequestActions.requestActionsToIgnore = [];
          serverRequestActions.successActions = [];
          serverRequestActions.dependentSuccessActions = [];
          serverRequestActions.errorActions = [];

        case 6:
          if (!true) {
            _context5.next = 30;
            break;
          }

          _context5.next = 9;
          return (0, _effects.take)(function (a) {
            return (0, _actions.isRequestAction)(a) || (0, _actions.isResponseAction)(a);
          });

        case 9:
          action = _context5.sent;

          if (!(0, _actions.isRequestAction)(action)) {
            _context5.next = 13;
            break;
          }

          index += action.meta && action.meta.dependentRequestsNumber !== undefined ? action.meta.dependentRequestsNumber + 1 : 1;
          return _context5.abrupt("continue", 6);

        case 13:
          if ((0, _actions.isSuccessAction)(action)) {
            _context5.next = 21;
            break;
          }

          serverRequestActions.errorActions.push(action);

          if (!finishOnFirstError) {
            _context5.next = 19;
            break;
          }

          _context5.next = 18;
          return (0, _effects.put)(_reduxSaga.END);

        case 18:
          return _context5.abrupt("return");

        case 19:
          _context5.next = 22;
          break;

        case 21:
          if (action.meta.isDependentRequest) {
            serverRequestActions.dependentSuccessActions.push(action);
          } else {
            serverRequestActions.successActions.push(action);
          }

        case 22:
          index -= action.meta.isDependentRequest ? 2 : 1;

          if (!(index === 0)) {
            _context5.next = 28;
            break;
          }

          serverRequestActions.requestActionsToIgnore = serverRequestActions.successActions.map(_actions.getRequestActionFromResponse).map(function (a) {
            return {
              type: a.type
            };
          });
          _context5.next = 27;
          return (0, _effects.put)(_reduxSaga.END);

        case 27:
          return _context5.abrupt("return");

        case 28:
          _context5.next = 6;
          break;

        case 30:
        case "end":
          return _context5.stop();
      }
    }
  }, _marked5, this);
}
